metaflow/datastore/content_addressed_store.py

```diff
@@ -145,38 +138,49 @@ def load_blobs(self, keys, force_raw=False):
         for k, path in zip(to_load, to_load_paths):
             # At this point, we either return the object as is (if raw) or
             # decode it according to the encoding version
-            result = load_results[path]
-            if force_raw or k.startswith('r_'):
+            result, meta = load_results[path]
+            if force_raw or (meta and meta.get('cas_raw', False)):
                 with result as r:
                     results[k] = r.read()
             else:
                 with result as r:
-                    magic, version = struct.unpack(
-                        self._info_format,
-                        r.read(struct.calcsize(self._info_format)))
-                    try:
-                        # We need to see if this is an artifact saved by this
-                        # version (or newer) of the datastore. Prior to this,
-                        # things were saved gzipped but with no magic string.
-                        if magic == self._magic:
-                            unpack_code = getattr(self, '_unpack_v%d' % version)
+                    if meta is None:
+                        # Backward compatible mode
+                        # We also need to deal with a short-lived backward
+                        # compatibility mode where the version was encoded
+                        # in the file itself
+                        magic = b'.MFBlob'
+                        # Pack format for header of non-raw blobs.
+                        # The integer corresponds to the version number of the encoding
+                        info_format = '<%dsi' % len(magic)
+                        file_magic, version = struct.unpack(
+                            info_format,
+                            r.read(struct.calcsize(info_format)))
+                        if file_magic == magic:
+                            unpack_code = getattr(
+                                self, '_unpack_v%d' % version, None)
                         else:
-                            # This is the backward compatible mode
                             r.seek(0)
+                            # This is the backward compatible mode
                             unpack_code = getattr(
                                 self, '_unpack_backward_compatible')
-                    except:
+                    else:
+                        version = meta.get('cas_version', -1)
+                        if version == -1:
+                            raise DataException(
+                                "Could not extract encoding version for %s" % path)
+                        unpack_code = getattr(self, '_unpack_v%d' % version, None)
+                    if unpack_code is None:
                         raise DataException(
                             "Unknown encoding version %d for %s -- the artifact "
                             "is either corrupt or you need to update Metaflow"
                             % (version, path))
```

Comment:
`version` may be undefined here. Just print the `path`.

Reply:
Fair point. As discussed. We can hopefully remove this as well (or at least simplify some of it).

Reply:
Done: I moved this code to the else part (only relevant part after refactoring to remove the version stored directly in the file.